core/src/main/java/io/ably/lib/realtime/Presence.java

```diff
@@ -1031,13 +1038,45 @@ synchronized void clear() {
                 residualMembers.clear();
         }
 
+        /**
+         * Spec: RTP17g
+         */
+        synchronized void reEnter() {
+            for (Map.Entry<String, PresenceMessage> entry: members.entrySet()) {
+                PresenceMessage member = entry.getValue();
+                member.action = PresenceMessage.Action.enter;
+                try {
+                    updatePresence(member, null);
+                } catch (AblyException e) {
+                    e.printStackTrace();
```

Comment:
Please use proper logging for this, instead of `e.printStackTrace()`. Currently, `java-sdk` is being used on different platforms, so it won't really work there as expected. e.g. android

Reply:
This is true for all `e.printStackTrace();` statements.

Reply:
You can check with @SimonWoolf on this, how to log errors. I think most likely errors are supposed to be passed back to the user who is calling the code so they can be handled gracefully, otherwise logged with proper logging level.

Reply:
RTP17e: `If the publish attempt fails for an automatic presence @ENTER@ (for example, by Ably rejecting it with a @NACK@), an @UPDATE@ event should be emitted on the channel with @resumed@ set to true and @reason@ set to an @ErrorInfo@ object with @code@ @91004@, a @message@ indicating that an automatic re-enter has failed and indicating the @clientId@, and @cause@ set to the reason for the enter failure. The error should also be logged at @warn@ level or higher.`

As usual you don't need to reinvent the wheel here, look at what the library already does in equivalent situations. Look a little further up this same file to endSyncAndEmitLeaves():
```
updatePresence(itemToSend, new CompletionListener() {
    @Override
    public void onSuccess() {
    }

    @Override
    public void onError(ErrorInfo reason) {
            /*
             * (RTP5c3)  If any of the automatic ENTER presence messages published
             * in RTP5c2 fail, then an UPDATE event should be emitted on the channel
             * with resumed set to true and reason set to an ErrorInfo object with error
             * code value 91004 and the error message string containing the message
             * received from Ably (if applicable), the code received from Ably
             * (if applicable) and the explicit or implicit client_id of the PresenceMessage
             */
        String errorString = String.format(Locale.ROOT, "Cannot automatically re-enter %s on channel %s (%s)",
                clientId, channel.name, reason.message);
        Log.e(TAG, errorString);
        channel.emitUpdate(new ErrorInfo(errorString, 91004), true);
    }
});
```

Reply:
Speaking of endSyncAndEmitLeaves, it... is still there, still implements the old auto-re-entry logic (RTP17c)? The conditions on when to auto-re-enter changed, but instead of changing the existing implementation you've just added the new one right next to the old, so now it's doing both?

Reply:
> there is not spec code anywhere. 

Yes, there is. In the same PR as the rest of the spec changes: https://github.com/ably/specification/pull/88/files#diff-d0e5a641ba4816766d38d6c0e7139df7502bed462dc676227057b4509cafad18R705 .

As you can see, `RTP17d` ("Automatic re-entry consists of, for each member of the RTP17 internal PresenceMap whose memberKey is not also a member of the normal PresenceMap...") was removed, and replaced with RTP17g, which says that reentry is unconditionally.

When this sort of requirement changes, you do need to use some common sense. Yes, it's _possible_ the library never implemented the old spec, but chances are it did, so look around and find the relevant implemention of the old spec, which you can modify to instead conform to the new spec.

> As I am still not sure what this thing does

You can see what it does by reading the code. It's not obfuscated. It implements the previous auto-re-entry spec.

Reply:
Ok I will take a look at it. When I meant no spec code anywhere I meant in the SDK code so finding things is harder. It will be changed soon as most of the code was documented with spec code.

Reply:
Thanks, I will take a look !

Reply:
Added more comments

Reply:
@sacOO7 I think this is it https://github.com/ably/ably-java/pull/842/commits/d022c3d5013d993f5c0bd49a0ec7a2c53e38fc46

Reply:
@SimonWoolf I have removed old implementation of re-enter as I already implemented the new logic in separate function reEnter() by spec RTP17g. 
You can check it here https://github.com/ably/ably-java/pull/842/commits/f0e09fb7d59c3581840db1f6fe891b316a33bfda